### DISCUSSION TITLE
New Benchmark::start() and Benchmark::stop() method

### DISCUSSION THREAD

**Discussion by khalyomede** [laravel/framework#45385] [Ideas]
With some collegues we wanted to know at which point our algorithm was taking a long time.

The business code is complex and uses Pipelines, each of the pipelines is suspected to take more or less time, and with `Benchmark` only supporting anonymous callback or array of callbacks it was impossible to split the pipeline easily.

We end up using `Log::debug()` and I think it's a super opportunity to bring a new idea to help Benchmark to be useful in even more scenarios.

Let's take a simplified example of the code I had to benchmark:

```php
namespace App\Http\Controllers;

use Illuminate\Pipeline\Pipeline;
use App\Filters\Filter1;
use App\Filters\Filter2;
use App\Filters\Filter3;
use App\Services\ServiceClass;

class ExampleController extends Controller
{
  public function index()
  {
    $items = app(Pipeline::class)
      ->send(ServiceClass::someMethod())
      ->through([
        Filter1::class,
        Filter2::class,
        Filter3::class,
      ])
      ->thenReturn();

    // ...
  }
}
```

My proposal is to create 2 new methods to be able to flexibly call the Benchmark class anywhere on the code, in any order, by tagging the laps (or not), and getting the result anywhere else, offering a lot of freedom and limiting the code changes at the minimum when we need to only draw a time report of a complex algorithm.

```php
namespace App\Http\Controllers;

use Illuminate\Pipeline\Pipeline;
use App\Filters\Filter1;
use App\Filters\Filter2;
use App\Filters\Filter3;
use App\Services\ServiceClass;

class ExampleController extends Controller
{
  public function index()
  {
    Benchmark::start(); // equivalent to Benchmark::start("default");

    $items = app(Pipeline::class)
      ->send(ServiceClass::someMethod())
      ->through([
        Filter1::class,
        Filter2::class,
        Filter3::class,
      ])
      ->thenReturn();

    Benchmark::stopAll(); // Could also call Benchmark::stop("name") for all 4
    Benchmark::dd();

    // Will dd
    // [
    //   "default" => "2312.0ms",
    //   "filter1" => "20.0ms",
    //   "filter2" => "1829.0ms",
    //   "filter3" => "458.0ms",
    // ]
  }
}
```

And this is what could be written in a filter for example:

```php
namespace App\Filters;

class Filter1
{
  public function handle(array $items)
  {
    Benchmark::start("filter1");

    // Some CPU intensive task

    Benchmark::stop("filter1");

    return $items:
  }
}
```

## Other ideas

- We could have access to `Benchmark::tags()` to know which tags exist and their state (started or stopped)
- `Benchmark::dd()` could automatically stop all remaining pending tags by convenience (so the code above do not really need to call `Benchmark::stopAll()` since dd would do it immediately)

**Comment by henzeb** [comment]
Laravel now a days comes with spatie/ray, which has that option. 
https://spatie.be/docs/ray/v1/usage/framework-agnostic-php#content-measuring-performance-and-memory-usage
